### PR TITLE
Remove `Falsy` from `TypeFacts.Truthy`

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/microsoft/TypeScript/issues/10564

⚠️ this only affects `strictNullChecks: false`

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -1205,7 +1205,7 @@ export const enum TypeFacts {
     FalseStrictFacts = BaseBooleanStrictFacts | Falsy,
     FalseFacts = BaseBooleanFacts,
     TrueStrictFacts = BaseBooleanStrictFacts | Truthy,
-    TrueFacts = BaseBooleanFacts | Truthy,
+    TrueFacts = (BaseBooleanFacts & ~Falsy) | Truthy,
```
Comment (Andarist) [inline src/compiler/checker.ts:1208]: if this is too broad of a change then perhaps the fix (if it's even desired today) could be moved closer to this `narrowType` callback:
https://github.dev/microsoft/TypeScript/blob/e936eb13d2900f21d79553c32a704307c7ad03dd/src/compiler/checker.ts#L27121

File: tests/baselines/reference/conditionalOperatorConditionIsBooleanType.types [hunk 2]
```diff
@@ -286,7 +286,7 @@ var resultIsString3 = null === undefined ? exprString1 : exprString2;
 var resultIsObject3 = true || false ? exprIsObject1 : exprIsObject2;
 >resultIsObject3 : Object
 >true || false ? exprIsObject1 : exprIsObject2 : Object
->true || false : boolean
+>true || false : true
```
Comment (RyanCavanaugh) [inline tests/baselines/reference/conditionalOperatorConditionIsBooleanType.types:289]: This is the wrong result under `strictNullChecks: false`

File: tests/baselines/reference/controlFlowWhileStatement.types [hunk 3]
```diff
@@ -158,7 +158,7 @@ function f() {
             break;
         }
         if (cond) {
->cond : true
+>cond : never
```
Comment (RyanCavanaugh) [inline tests/baselines/reference/controlFlowWhileStatement.types:161]: !

File: tests/baselines/reference/conditionalOperatorConditionIsBooleanType.types (hunk 2 above)
Comment (Andarist) [reply tests/baselines/reference/conditionalOperatorConditionIsBooleanType.types:289]: Could u ELI5 how to reason about `strictNullChecks: false` in situations like this?

Comment (RyanCavanaugh) [reply tests/baselines/reference/conditionalOperatorConditionIsBooleanType.types:289]: Any value can be falsy due to being `null` / `undefined`

Comment (Andarist) [reply tests/baselines/reference/conditionalOperatorConditionIsBooleanType.types:289]: I know that from the assignability PoV but that doesnt help me explain your expected result here. Shouldnt intrinsic value like `true` be non-falsy? 

Where is the line if the referenced issue got labeled as a bug?

Comment (RyanCavanaugh) [reply tests/baselines/reference/conditionalOperatorConditionIsBooleanType.types:289]: A value of type `true` might be inhabited by `null`. Obviously the expression `true` itself isn't, but that's not really under analysis

Comment (Andarist) [reply tests/baselines/reference/conditionalOperatorConditionIsBooleanType.types:289]: If the compiler has to always account for the fact that the value might be inhabited by `null` then wouldn't it mean that one of the examples from the referenced issue shouldn't typecheck (and it does typecheck today)?
```ts
type Result = { success: true }
            | { success: false, error: string }

function handleError2(res: Result) {
    if (res.success !== false) {
        return;
    }

    res.error; // OK, but shouldn't be?
}
```

I always thought that `strictNullChecks: false` means that things can be `null`/`undefined` at runtime but that the type-checker is meant to ignore this fact completely.

Comment (Andarist) [reply tests/baselines/reference/conditionalOperatorConditionIsBooleanType.types]: Either way - I pushed out an improved (but not cleaned up, code-wise) version of the fix for the linked issue. Please take a look if those semantics look any better.

Comment (fatcerberus) [reply tests/baselines/reference/conditionalOperatorConditionIsBooleanType.types]: > I always thought that strictNullChecks: false means that things can be null/undefined at runtime but that the type-checker is meant to ignore this fact completely.

This is a point of discussion that comes up sometimes on Discord - `strictNullChecks: false` is basically a completely different type checker.  The flag does a lot more than its name would suggest, changing inferences, narrowing, etc. in sometimes very unexpected ways, and all of it is more or less by design.

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this
@typescript-bot test this
@typescript-bot test top300
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this

**Comment by sandersn** [comment]
To help with PR housekeeping, I'm going to close this PR since it's pretty old now.
strictNullChecks: false is already a mess of inconsistent semantics that fewer and fewer people use. We definitely shouldn't change them.
